### PR TITLE
[Basic Editor] -  fully integrated the basic editor 

### PR DESCRIPTION
**Pull request by gastonche**
I have successfully integrated the  basic UI editor (first part of #5 ) for the library and also done the following:

- [X] changed the whole build process to use webpack instead of browserfy

- [X] succeeded to make the library a js only library, such that the user only needs to include the js file and it generates the css

- [X] corrected the travis build to enable linting before install and tests after builds effectively:  fixing issue #7 

- [X] Enabled json-schema changing with a changeSchema method for developers #3 

- [X] added more tests for advanced module and corrected failing tests

### CODE REVIEW

File: README.rst [hunk 1]
```diff
@@ -62,20 +62,17 @@ Development Setup
 To setup this project for development do the following
 
 1. Fork the repository.
-2. Clone the repository locally using ``git clone``.
+2. Clone the repository locally using ``git clone https://github.com/xxxxxx/netjsonconfig-editor.js.git``.
```
Comment (nemesisdesign) [inline README.rst]: use `https://github.com/<your_fork>/netjsonconfig-editor.js.git` for consistency with other OpenWISP repositories

File: src/css/basic.css [hunk 2]
```diff
@@ -0,0 +1,50 @@
+.padding{
+	padding: 20px;
```
Comment (nemesisdesign) [inline src/css/basic.css]: whitespace looks wrong

File: src/css/basic.css [hunk 3]
```diff
@@ -0,0 +1,50 @@
...
+}
+
+.basic-header{
+	min-height: 50px;
```
Comment (nemesisdesign) [inline src/css/basic.css]: whitespace looks wrong

File: src/css/basic.css [hunk 4]
```diff
@@ -0,0 +1,50 @@
...
+}
+.basic-header .advanced-button-container{
+	width: 20%;
+	display: inline-block;
+	text-align: left;
+
+}
+
+.advanced-button{
+	background: #5e6569;
```
Comment (nemesisdesign) [inline src/css/basic.css]: whitespace looks wrong

File: src/js/basic/body.js [hunk 5]
```diff
@@ -0,0 +1,87 @@
...
+	error: PropTypes.object,
+	theme: PropTypes.array
+}
+
+BasicBody.propTypes = {
+	schema: PropTypes.object,
+	onChange: PropTypes.func,
+	data: PropTypes.object,
+
+
```
Comment (nemesisdesign) [inline src/js/basic/body.js]: remove unnecessary blank lines

File: src/js/basic/index.js [hunk 6]
```diff
@@ -0,0 +1,21 @@
+import React from 'react';
+import ReactDOM from 'react-dom';
+import App from './app';
+
+class BasicEditor {
+
```
Comment (nemesisdesign) [inline src/js/basic/index.js]: remove unnecessary blank line

File: src/js/basic/index.js [hunk 7]
```diff
@@ -0,0 +1,21 @@
...
+	constructor({target, schema, data, helpText, validate, onChange, jsonError, container, swapOut}){
+		this.render({target, schema, data, helpText, validate, onChange, jsonError, container, swapOut});
```
Comment (nemesisdesign) [inline src/js/basic/index.js]: can we reformat this to something like the following?

```javascript
constructor({target, schema, data, helpText, validate,
             onChange, jsonError, container, swapOut}){
    this.render({target, schema, data, helpText, validate,
                 onChange, jsonError, container, swapOut});
    // ecc...
}
```

File: examples/index.html [hunk 8]
```diff
@@ -19,7 +18,7 @@
 					man: "hello"
 				}, 
 				onChange: function(){
-					alert('hello');
+                    var worked = true;
```
Comment (nemesisdesign) [inline examples/index.html]: this line doesn't look indented correctly, what do you think?

File: src/js/advanced/index.js [hunk 9]
```diff
@@ -1,118 +1,176 @@
...
-};
+  /**
+   * Represents an instance of the Advanced Editor.
+   * @constructor
+   * @param {object} props - propts containing the configs for the editor
+   */
+  constructor({target, schema, data, helpText, validate,
+            onChange, jsonErrorMessage, swapOut}) {
+    this.swapOut = swapOut;
+        // set message to be used to alertwrong json
```
Comment (nemesisdesign) [inline src/js/advanced/index.js]: this line?

File: src/js/advanced/index.js [hunk 10]
```diff
@@ -1,118 +1,176 @@
...
+    this.setJsonErrorMessage(jsonErrorMessage);
+        // disable validation if not required
```
Comment (nemesisdesign) [inline src/js/advanced/index.js]: this line?

File: src/js/advanced/index.js [hunk 11]
```diff
@@ -1,118 +1,176 @@
...
+    this.validate = validate;
+
+    schema = validate ? schema : {};
+        // build options for advanced mode
```
Comment (nemesisdesign) [inline src/js/advanced/index.js]: this line?

File: src/js/advanced/index.js [hunk 12]
```diff
@@ -1,118 +1,176 @@
...
+    const ajv = new Ajv({
+      meta: false, // optional, to prevent adding draft-06 meta-schema
+      extendRefs: true, // optional, current default is to 'fail'
+      unknownFormats: 'ignore',  // optional, current default is true (fail)
+    });
+
+    ajv.addMetaSchema(metaSchema);
+    ajv._opts.defaultMeta = metaSchema.id;
+
+        // optional, using unversioned URI is out of spec, see https://github.com/json-schema-org/json-schema-spec/issues/216
```
Comment (nemesisdesign) [inline src/js/advanced/index.js]: this line?

File: src/js/advanced/index.js [hunk 13]
```diff
@@ -1,118 +1,176 @@
...
+    ajv._refs['http://json-schema.org/schema'] = 'http://json-schema.org/draft-04/schema';
+
+        // Optionally you can also disable keywords defined in draft-06
```
Comment (nemesisdesign) [inline src/js/advanced/index.js]: this line?

File: src/js/advanced/index.js [hunk 14]
```diff
@@ -1,118 +1,176 @@
...
+    ajv.removeKeyword('propertyNames');
+    ajv.removeKeyword('contains');
+    ajv.removeKeyword('const');
+
+        // ajv fixes end here
```
Comment (nemesisdesign) [inline src/js/advanced/index.js]: this line?

File: src/js/advanced/index.js [hunk 15]
```diff
@@ -1,118 +1,176 @@
...
+
+        // create advanced mode element with jQuery and insert into the DOM
```
Comment (nemesisdesign) [inline src/js/advanced/index.js]: this line?

File: src/js/advanced/index.js [hunk 16]
```diff
@@ -1,118 +1,176 @@
...
+    this.element = $('<div class=\'advanced-mode full-screen\'></div>');
+    this.element.appendTo($(target));
+
+    this.editor = new JsonEditor(this.element[0], {...options, ajv}, data);
+    this.editor.aceEditor.setOptions({
+      fontSize: 14,
+      showInvisibles: true,
+    });
+
+        // remove powered by ace link
```
Comment (nemesisdesign) [inline src/js/advanced/index.js]: this line?

File: src/js/advanced/index.js [hunk 17]
```diff
@@ -1,118 +1,176 @@
...
+    this.element.find('.jsoneditor-menu a').remove();
+
+    const that = this;
+
+        // add controls to the editor header
```
Comment (nemesisdesign) [inline src/js/advanced/index.js]: this line?

File: src/js/index.js [hunk 18]
```diff
@@ -1,67 +1,168 @@
...
+
+    validate = !validate? false: true;
+
+    let onChangeCb = onChange? onChange: () => {};
+
+    jsonError = jsonError? jsonError: 'Json entered is invalid';
+
+    data = data? data: {};
+
+
```
Comment (nemesisdesign) [inline src/js/index.js]: can you remove some unnecessary blank lines to make the code more compact please?

File: src/js/index.js [hunk 19]
```diff
@@ -1,67 +1,168 @@
...
+      this.changed(onChangeCb, data);
+    };
+
+    helpText = helpText? helpText:
+          `Want learn to use the advanced mode? Consult the
+          <a href='http://netjsonconfig.openwisp.org/en/stable/general/basics.html'
+             target='_blank'>netjsonconfig documentation
+          </a>.`;
+
+    this.props = {target, helpText,
```
Comment (nemesisdesign) [inline src/js/index.js]: what about:

```javascript
this.props = {target, helpText, data, schema,
              validate, onChange, jsonError};
```

File: src/js/index.js [hunk 20]
```diff
@@ -1,67 +1,168 @@
...
+    this.setJson(data);
+  }
+  /**
+   * Method used to render the instance of the editor.
+   * @param {Object} props
+   */
+  render({helpText, data, schema, validate, onChange, jsonError}) {
+    this.container = $(`
+        <div class='netjsoneditor-config'></div>
+      `);
```
Comment (nemesisdesign) [inline src/js/index.js]: this line

File: src/js/index.js [hunk 21]
```diff
@@ -1,67 +1,168 @@
...
+    this.container.insertBefore($(this.targetElement));
+
+    this.initAdvancedEditor({target: this.container, helpText,
+      data, schema, validate, onChange, jsonError});
```
Comment (nemesisdesign) [inline src/js/index.js]: what about reformatting this line too as one of my previous suggestions?

File: src/js/index.js [hunk 22]
```diff
@@ -1,67 +1,168 @@
...
+    this.initBasicEditor({target: this.container,
+      helpText, data, schema, validate, onChange, jsonError});
```
Comment (nemesisdesign) [inline src/js/index.js]: what about reformatting this line too as one of my previous suggestions?

File: src/js/index.js [hunk 23]
```diff
@@ -1,67 +1,168 @@
...
+    this.basicEditor.setData(json);
+  }
+  /**
+   * getter method for the json schema being used within the editor.
+   * @return {Object} schema
+   */
+  get schema() {
+    return this.hschema;
+  }
+
```
Comment (nemesisdesign) [inline src/js/index.js]: remove blank line

File: test/tests.js [hunk 24]
```diff
@@ -1,39 +1,134 @@
...
+                it("Should have the basicEditor visible on startup", function() {
+                    expect(editor.basicEditor.container.css("display")).to.not.be.equals('none'); 
+                });
+
+                it("Should have the advancedEditor hidden on startup", function() {
+                    expect(editor.advancedEditor.element.css("display")).to.be.equals('none'); 
+                });
+            });
+
+
```
Comment (nemesisdesign) [inline test/tests.js]: remove blank line

File: test/tests.js [hunk 25]
```diff
@@ -1,39 +1,134 @@
...
-        it('should have a text property equals to the stringified json property', function() {
-         expect(editor.text).to.equals(JSON.stringify(editor.json));
-     });
+    makeSuite('initialised with empty object', {}, function(){
+        tests();
+    });
 
+    opts = {
+                data: {
+                   "files": [
```
Comment (nemesisdesign) [inline test/tests.js]: indentation looks wrong in this block

File: test/tests.js [hunk 26]
```diff
@@ -1,39 +1,134 @@
...
+            var worked = true;
+        }
     };
 
-    editor = new netjsonEditor({});
-    describe('initialised with empty object', tests);
+    makeSuite('initialised with '+JSON.stringify(opts)+' object', opts, function(){
+        tests();
+    });
+
```
Comment (nemesisdesign) [inline test/tests.js]: remove blank line

**Comment by gastonche** [comment]
I checked the reason why tests fail on your local env, and it's because you need to have run ``npm run build`` before you can run ``npm test``, that's why you get the error you are getting.

**Comment by nemesisdesign** [comment]
@gastonche you were right, after running both commands it went well, did you write this in the README?

**Pull request merged**